File: app/typescript/src/application/port/in/FinancialAnalysisService.ts
// src/application/port/in/FinancialAnalysisService.ts

import type { JournalRepository } from '../out/JournalRepository'
import { FinancialData } from '../../../domain/model/financial-analysis/FinancialData'
import { FinancialRatioAnalyzer } from '../../../domain/model/financial-analysis/FinancialRatioAnalyzer'
import type {
  FinancialAnalysisUseCase,
  FinancialAnalysisResult,
  ComparativeAnalysisResult
} from './FinancialAnalysisUseCase'

export class FinancialAnalysisService implements FinancialAnalysisUseCase {
  private readonly analyzer: FinancialRatioAnalyzer

  constructor(private readonly journalRepository: JournalRepository) {
    this.analyzer = new FinancialRatioAnalyzer()
  }

  async analyzeByFiscalYear(fiscalYear: number): Promise<FinancialAnalysisResult> {
    // 会計年度の開始日と終了日を計算（日本の会計年度：4月1日〜翌年3月31日）
    const startDate = new Date(fiscalYear - 1, 3, 1) // 4月1日（月は0始まり）
    const endDate = new Date(fiscalYear, 2, 31) // 3月31日

    // 指定された会計年度の仕訳を取得
    const journals = await this.journalRepository.findByDateRange(startDate, endDate)

    if (journals.length === 0) {
      throw new Error(`会計年度 ${fiscalYear} の仕訳データが見つかりません`)
    }

    // 仕訳明細から勘定科目別の金額を集計
    const entries = journals.flatMap((journal) =>
      journal.details.flatMap((detail) =>
        detail.items.map((item) => ({
          accountCode: item.accountCode,
          debitAmount: item.debitOrCredit === '借方' ? item.amount : 0,
          creditAmount: item.debitOrCredit === '貸方' ? item.amount : 0
        }))
      )
    )

    // 勘定科目ごとに集計
    const accountMap = new Map<string, { debitAmount: number; creditAmount: number }>()
    for (const entry of entries) {
      const existing = accountMap.get(entry.accountCode) || { debitAmount: 0, creditAmount: 0 }
      existing.debitAmount += entry.debitAmount
      existing.creditAmount += entry.creditAmount
      accountMap.set(entry.accountCode, existing)
    }

    // Map を配列に変換
    const aggregatedEntries = Array.from(accountMap.entries()).map(([accountCode, amounts]) => ({
      accountCode,
      debitAmount: amounts.debitAmount,
      creditAmount: amounts.creditAmount
    }))

    // 財務データを生成
    const financialData = FinancialData.fromJournalEntries(fiscalYear, aggregatedEntries)

    // 財務分析を実行
    const ratios = this.analyzer.analyze(financialData)

    return {
      fiscalYear,
      financialData: {
        sales: financialData.sales,
        operatingProfit: financialData.operatingProfit,
        totalAssets: financialData.totalAssets,
        tangibleFixedAssets: financialData.tangibleFixedAssets,
        currentAssets: financialData.currentAssets,
        currentLiabilities: financialData.currentLiabilities,
        quickAssets: financialData.quickAssets,
        equity: financialData.equity
      },
      ratios
    }
  }

  async compareMultiplePeriods(fiscalYears: number[]): Promise<ComparativeAnalysisResult> {
    // 各期間の分析を実行
    const periods = await Promise.all(
      fiscalYears.map((fiscalYear) => this.analyzeByFiscalYear(fiscalYear))
    )

    // 最初の期間と最後の期間の差分を計算
    const first = periods[0]
    const last = periods[periods.length - 1]

    const trends = {
      profitability: {
        operatingProfitMarginChange:
          last.ratios.profitability.operatingProfitMargin -
          first.ratios.profitability.operatingProfitMargin
      },
      efficiency: {
        totalAssetTurnoverChange:
          last.ratios.efficiency.totalAssetTurnover - first.ratios.efficiency.totalAssetTurnover
      },
      safety: {
        equityRatioChange: last.ratios.safety.equityRatio - first.ratios.safety.equityRatio
      }
    }

    return {
      periods,
      trends
    }
  }
}

File: app/java/app/src/main/java/com/example/accounting/application/service/management/FinancialAnalysisService.java
package com.example.accounting.application.service.management;

import com.example.accounting.application.port.in.*;
import com.example.accounting.application.port.out.DailyBalanceRepository;
import com.example.accounting.domain.model.management.AnalysisResult;
import com.example.accounting.domain.model.management.FinancialData;
import com.example.accounting.domain.model.management.FinancialRatioAnalyzer;
import com.example.accounting.domain.model.financial.DailyBalance;
import com.example.accounting.infrastructure.in.web.dto.management.*;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;

/**
 * 財務分析サービス
 *
 * 日次残高データから財務データを集計し、財務指標を計算します。
 */
@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class FinancialAnalysisService implements FinancialAnalysisUseCase {

    private final DailyBalanceRepository dailyBalanceRepository;
    private final FinancialRatioAnalyzer analyzer = new FinancialRatioAnalyzer();

    @Override
    public FinancialAnalysisResult analyzeByFiscalYear(int fiscalYear) {
        // 会計年度の期末日を計算（3月31日）
        LocalDate fiscalYearEnd = LocalDate.of(fiscalYear + 1, 3, 31);

        // 期末日の日次残高データを取得
        List<DailyBalance> balances = dailyBalanceRepository.findByDate(fiscalYearEnd);

        if (balances.isEmpty()) {
            throw new IllegalArgumentException(
                String.format("会計年度 %d の財務データが見つかりません", fiscalYear)
            );
        }

        // 日次残高から仕訳明細データに変換
        List<FinancialData.JournalEntryData> entries = balances.stream()
            .map(balance -> new FinancialData.JournalEntryData(
                balance.getAccountCode(),
                balance.getDebitAmount(),
                balance.getCreditAmount()
            ))
            .toList();

        // 財務データを生成
        FinancialData financialData = FinancialData.fromJournalEntries(fiscalYear, entries);

        // 財務分析を実行
        AnalysisResult ratios = analyzer.analyze(financialData);

        // 結果を DTO に変換
        return new FinancialAnalysisResult(
            fiscalYear,
            toFinancialDataDto(financialData),
            toRatiosDto(ratios)
        );
    }

    @Override
    public ComparativeAnalysisResult compareMultiplePeriods(List<Integer> fiscalYears) {
        // 各期間の分析を実行
        List<FinancialAnalysisResult> periods = fiscalYears.stream()
            .map(this::analyzeByFiscalYear)
            .toList();

        if (periods.size() < 2) {
            return new ComparativeAnalysisResult(
                periods,
                new TrendsDto(BigDecimal.ZERO, BigDecimal.ZERO, BigDecimal.ZERO)
            );
        }

        // 最初の期間と最後の期間の差分を計算
        FinancialAnalysisResult first = periods.get(0);
        FinancialAnalysisResult last = periods.get(periods.size() - 1);

        TrendsDto trends = new TrendsDto(
            last.ratios().profitability().operatingProfitMargin()
                .subtract(first.ratios().profitability().operatingProfitMargin()),
            last.ratios().efficiency().totalAssetTurnover()
                .subtract(first.ratios().efficiency().totalAssetTurnover()),
            last.ratios().safety().equityRatio()
                .subtract(first.ratios().safety().equityRatio())
        );

        return new ComparativeAnalysisResult(periods, trends);
    }

    /**
     * FinancialData を DTO に変換
     */
    private FinancialDataDto toFinancialDataDto(FinancialData data) {
        return new FinancialDataDto(
            data.sales(),
            data.operatingProfit(),
            data.totalAssets(),
            data.tangibleFixedAssets(),
            data.currentAssets(),
            data.currentLiabilities(),
            data.quickAssets(),
            data.equity()
        );
    }

    /**
     * AnalysisResult を DTO に変換
     */
    private RatiosDto toRatiosDto(AnalysisResult ratios) {
        return new RatiosDto(
            new ProfitabilityDto(ratios.profitability().operatingProfitMargin()),
            new EfficiencyDto(
                ratios.efficiency().totalAssetTurnover(),
                ratios.efficiency().tangibleFixedAssetTurnover()
            ),
            new SafetyDto(
                ratios.safety().currentRatio(),
                ratios.safety().quickRatio(),
                ratios.safety().equityRatio()
            )
        );
    }
}
